"Module not found" errors when using knex with webpack

I am trying to use knex in a project I'm working on. The problem I'm having is that when I try to run webpack it is tracing the requires and attempting to pull in drivers/dialects that I'm not using and don't have installed.

Any thoughts on how this can be solved?

`````` log
ERROR in ./~/mariasql/lib/Client.js
Module not found: Error: Cannot resolve 'file' or 'directory' ../build/Debug/sqlclient in ~/project/node_modules/mariasql/lib
 @ ./~/mariasql/lib/Client.js 17:10-45

ERROR in ./~/knex/lib/dialects/sqlite3/index.js
Module not found: Error: Cannot resolve module 'sqlite3' in ~/project/node_modules/knex/lib/dialects/sqlite3
 @ ./~/knex/lib/dialects/sqlite3/index.js 33:11-29

ERROR in ./~/knex/lib/dialects/mysql2/index.js
Module not found: Error: Cannot resolve module 'mysql2' in ~/project/node_modules/knex/lib/dialects/mysql2
 @ ./~/knex/lib/dialects/mysql2/index.js 33:11-28

ERROR in ./~/knex/lib/dialects/mysql/index.js
Module not found: Error: Cannot resolve module 'mysql' in ~/project/node_modules/knex/lib/dialects/mysql
 @ ./~/knex/lib/dialects/mysql/index.js 35:11-27

ERROR in ./~/knex/lib/dialects/oracle/index.js
Module not found: Error: Cannot resolve module 'oracle' in ~/project/node_modules/knex/lib/dialects/oracle
 @ ./~/knex/lib/dialects/oracle/index.js 40:11-28

ERROR in ./~/knex/lib/dialects/postgres/index.js
Module not found: Error: Cannot resolve module 'pg' in ~/project/node_modules/knex/lib/dialects/postgres
 @ ./~/knex/lib/dialects/postgres/index.js 46:11-24

ERROR in ./~/knex/lib/dialects/postgres/index.js
Module not found: Error: Cannot resolve module 'pg-query-stream' in ~/project/node_modules/knex/lib/dialects/postgres
 @ ./~/knex/lib/dialects/postgres/index.js 132:50-76

ERROR in ./~/knex/lib/dialects/strong-oracle/index.js
Module not found: Error: Cannot resolve module 'strong-oracle' in ~/project/node_modules/knex/lib/dialects/strong-oracle
 @ ./~/knex/lib/dialects/strong-oracle/index.js 15:9-33```
``````

--- a/index.html
+++ b/index.html
@@ -2224,18 +2224,27 @@ knex('users')
   .orWhere(knex.raw('status <> ?', [1]))
   .groupBy('status')
 </pre>
+
+    <p>
+      Positional bindings <tt>?</tt> are interpreted as values and <tt>??</tt> are interpreted as identifiers.
+    </p>
+
 <pre class="display">
-// positional bindings ? is interpret as value and ?? as identifier
 knex('users').where(knex.raw('?? = ?', ['user.name', 1]))
 </pre>
 
+    <p>
+      Named bindings such as <tt>:name</tt> are interpreted as values and <tt>:name:</tt> interpreted as identifiers.
+      Named bindings are processed so long as the value is anything other than <tt>undefined</tt>.
+    </p>
+
 <pre class="display">
-// named bindings :name is interpret as value and :name: as identifier
 knex('users')
-  .where(knex.raw(':name: = :thisGuy or :name: = :otherGuy', {
+  .where(knex.raw(':name: = :thisGuy or :name: = :otherGuy or :name: = :undefinedBinding', {
     name: 'users.name',
     thisGuy: 'Bob',
-    otherGuy: 'Jay'
+    otherGuy: 'Jay',
+    undefinedBinding: undefined
   }))
 </pre>
 

--- a/src/raw.js
+++ b/src/raw.js
@@ -127,7 +127,7 @@ function replaceKeyBindings(raw) {
   var client   = raw.client
   var sql      = raw.sql, bindings = []
 
-  var regex = new RegExp('(^|\\s)(\\:\\w+\\:?)', 'g')
+  var regex = new RegExp('(\\:\\w+\\:?)', 'g')
   sql = raw.sql.replace(regex, function(full) {
     var key = full.trim();
     var isIdentifier = key[key.length - 1] === ':'

--- a/src/raw.js
+++ b/src/raw.js
@@ -128,11 +128,15 @@ function replaceKeyBindings(raw) {
   var sql      = raw.sql, bindings = []
 
   var regex = new RegExp('(\\:\\w+\\:?)', 'g')
+  var emptyBindings = [];
   sql = raw.sql.replace(regex, function(full) {
     var key = full.trim();
     var isIdentifier = key[key.length - 1] === ':'
     var value = isIdentifier ? values[key.slice(1, -1)] : values[key.slice(1)]
-    if (value === undefined) return ''
+    if (value === undefined) {
+      emptyBindings.push(full);
+      return full;
+    }
     if (value && typeof value.toSQL === 'function') {
       var bindingSQL = value.toSQL()
       if (bindingSQL.bindings !== undefined) {
@@ -146,6 +150,10 @@ function replaceKeyBindings(raw) {
     bindings.push(value)
     return full.replace(key, '?')
   })
+
+  emptyBindings.map(function(binding) {
+    sql = sql.replace(new RegExp(` ${binding}|${binding} |${binding}`, 'g'), '');
+  });
 
   return {
     method: 'raw',

--- a/src/raw.js
+++ b/src/raw.js
@@ -128,13 +128,11 @@ function replaceKeyBindings(raw) {
   var sql      = raw.sql, bindings = []
 
   var regex = new RegExp('(\\:\\w+\\:?)', 'g')
-  var emptyBindings = [];
   sql = raw.sql.replace(regex, function(full) {
     var key = full.trim();
     var isIdentifier = key[key.length - 1] === ':'
     var value = isIdentifier ? values[key.slice(1, -1)] : values[key.slice(1)]
     if (value === undefined) {
-      emptyBindings.push(full);
       return full;
     }
     if (value && typeof value.toSQL === 'function') {
@@ -150,10 +148,6 @@ function replaceKeyBindings(raw) {
     bindings.push(value)
     return full.replace(key, '?')
   })
-
-  emptyBindings.map(function(binding) {
-    sql = sql.replace(new RegExp(` ${binding}|${binding} |${binding}`, 'g'), '');
-  });
 
   return {
     method: 'raw',

--- a/test/tape/raw.js
+++ b/test/tape/raw.js
@@ -71,18 +71,18 @@ test('allows for options in raw queries, #605', function(t) {
   })
 })
 
-test('raw query strings with keys replace values', function(t) {
-
+test('undefined named bindings are ignored', function(t) {
+  
   t.plan(2)
+  
+  t.equal(raw('select :item from :place', {}).toSQL().sql, 'select :item from :place')
 
-  t.equal(raw('select :item from :place', {}).toSQL().sql, 'select from')
-
-  t.equal(raw('select :item :cool 2 from :place', {}).toSQL().sql, 'select 2 from')
+  t.equal(raw('select :item :cool 2 from :place', {item: 'col1'}).toSQL().sql, 'select ? :cool 2 from :place')
 
 })
 
 test('raw bindings are optional, #853', function(t) {
-
+  
   t.plan(2)
 
   var sql = raw('select * from ? where id=?', [raw('foo'), 4]).toSQL()

--- a/test/unit/query/builder.js
+++ b/test/unit/query/builder.js
@@ -3207,4 +3207,47 @@ describe("QueryBuilder", function() {
       }
     });
   });
+
+  it('#1228 Named bindings', function() {
+    testsql(qb().select('*').from('users').whereIn('id', raw('select (:test)', {test: [1,2,3]})), {
+      mysql: {
+        sql: 'select * from `users` where `id` in (select (?))',
+        bindings: [[1,2,3]]
+      },
+      mssql: {
+        sql: 'select * from [users] where [id] in (select (?))',
+        bindings: [[1,2,3]]
+      },
+      default: {
+        sql: 'select * from "users" where "id" in (select (?))',
+        bindings: [[1,2,3]]
+      }
+    });
+
+
+    var namedBindings = {
+      name:     'users.name',
+      thisGuy:  'Bob',
+      otherGuy: 'Jay'
+    };
+    //Had to do it this way as the 'raw' statement's .toQuery is called before testsql, meaning mssql and other dialects would always get the output of qb() default client
+    //as MySQL, which means testing the query per dialect won't work. [users].[name] would be `users`.`name` for mssql which is incorrect.
+    var mssql = clients.mssql;
+    var mysql = clients.mysql;
+    var defaultClient = clients.default;
+
+    var mssqlQb = mssql.queryBuilder().select('*').from('users').where(mssql.raw(':name: = :thisGuy or :name: = :otherGuy', namedBindings)).toSQL();
+    var mysqlQb = mysql.queryBuilder().select('*').from('users').where(mysql.raw(':name: = :thisGuy or :name: = :otherGuy', namedBindings)).toSQL();
+    var defaultQb = defaultClient.queryBuilder().select('*').from('users').where(defaultClient.raw(':name: = :thisGuy or :name: = :otherGuy', namedBindings)).toSQL();
+
+    expect(mssqlQb.sql).to.equal('select * from [users] where [users].[name] = ? or [users].[name] = ?');
+    expect(mssqlQb.bindings).to.deep.equal(['Bob', 'Jay']);
+
+    expect(mysqlQb.sql).to.equal('select * from `users` where `users`.`name` = ? or `users`.`name` = ?');
+    expect(mysqlQb.bindings).to.deep.equal(['Bob', 'Jay']);
+
+    expect(defaultQb.sql).to.equal('select * from "users" where "users"."name" = ? or "users"."name" = ?');
+    expect(defaultQb.bindings).to.deep.equal(['Bob', 'Jay']);
+
+  });
 });

